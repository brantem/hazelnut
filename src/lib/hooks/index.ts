export * from 'lib/hooks/useSearch';
export * from 'lib/hooks/useModal';
