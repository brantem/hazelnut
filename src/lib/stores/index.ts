import { useEffect } from 'react';

import { itemsStore } from 'lib/stores/items';
import { groupsStore } from 'lib/stores/groups';
import { routinesStore } from 'lib/stores/routines';
import { historiesStore } from 'lib/stores/histories';
import storage from 'lib/stores/storage';

/* c8 ignore start */
export const useLoadStore = () => {
  useEffect(() => {
    (async () => {
      const [items, groups, routines, histories] = await Promise.all([
        storage.getAll('items'),
        storage.getAll('groups'),
        storage.getAll('routines'),
        storage.getAll('histories'),
      ]);

      itemsStore.setState({ items: items.sort((a, b) => a.createdAt - b.createdAt), isReady: true });
      groupsStore.setState({ groups: groups.sort((a, b) => a.createdAt - b.createdAt), isReady: true });
      routinesStore.setState({ routines, isReady: true });
      historiesStore.setState({ histories, selectedDate: localStorage.getItem('history-selected-date') });
    })();
  }, []);
};
/* c8 ignore stop */

export * from 'lib/stores/items';
export * from 'lib/stores/groups';
export * from 'lib/stores/routines';
export * from 'lib/stores/search';
export * from 'lib/stores/histories';
export * from 'lib/stores/modal';
