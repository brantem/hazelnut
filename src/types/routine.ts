import { Day } from 'types/shared';

export type Routine = {
  id: string;
  title: string;
  color: string;
  days?: Day[];
  recurrence?: {
    startAt: number;
    interval: number;
  };
  time: string | null;
  itemIds: string[];
  minimized: boolean;
  createdAt: number;
};
