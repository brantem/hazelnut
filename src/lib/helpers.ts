import days from 'data/days';
import { Day } from 'types/shared';

export const getMinutesFromTime = (time: string) => {
  const [hour, minute] = time.split(':');
  return parseInt(hour) * 60 + parseInt(minute);
};

export const getCurrentDay = (): Day => [...days.slice(6), ...days.slice(0, -1)][new Date().getDay()];
