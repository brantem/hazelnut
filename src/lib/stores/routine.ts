import create from 'zustand';

import { Routine } from 'types/routine';

type RoutineState = {
  routine: Routine | null;
  clear: () => void;

  isSaveOpen: boolean;
  showSave: (routine?: Routine | null) => void;

  isSaveItemsOpen: boolean;
  showSaveItems: (routine?: Routine | null) => void;

  isSettingsOpen: boolean;
  showSettings: (routine: Routine) => void;

  hide: () => void;
};

export const useRoutineStore = create<RoutineState>()((set) => ({
  routine: null,
  clear: () => set({ routine: null }),

  isSaveOpen: false,
  showSave: (routine = null) => set((state) => ({ routine: routine || state.routine, isSaveOpen: true })),

  isSaveItemsOpen: false,
  showSaveItems: (routine) => set({ routine, isSaveItemsOpen: true }),

  isSettingsOpen: false,
  showSettings: (routine) => set({ routine, isSettingsOpen: true }),

  hide: () => set({ isSaveOpen: false, isSaveItemsOpen: false, isSettingsOpen: false }),
}));
