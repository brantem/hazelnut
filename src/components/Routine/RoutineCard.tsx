import { EllipsisVerticalIcon } from '@heroicons/react/20/solid';

import { useRoutineStore } from 'lib/stores';
import type { Routine } from 'types/routine';

type RoutineProps = {
  routine: Routine;
  showAction?: boolean;
};

const RoutineCard = ({ routine, showAction }: RoutineProps) => {
  const { showSaveItems, showSettings } = useRoutineStore();

  return (
    <div className={`px-4 py-3 bg-${routine.color}-50`} data-testid="routine-card">
      <div className="flex justify-between items-center space-x-3 h-7">
        <div className="flex items-center space-x-3 max-w-[440px]">
          <h3 className={`uppercase text-sm font-semibold text-${routine.color}-600 truncate`}>{routine.title} </h3>
          <span className={`font-medium text-sm text-${routine.color}-400 ml-2 tabular-nums flex-shrink-0`}>
            {routine.time}
          </span>
        </div>

        {showAction && (
          <div className="flex items-center space-x-1 flex-shrink-0">
            <button
              className={`px-2 py-1 text-sm rounded-md hover:bg-${routine.color}-100 flex-shrink-0`}
              onClick={() => showSaveItems(routine)}
              data-testid="routine-card-items"
            >
              Items
            </button>

            <button
              className={`p-1 rounded-md hover:bg-${routine.color}-100`}
              onClick={() => showSettings(routine)}
              data-testid="routine-card-settings"
            >
              <EllipsisVerticalIcon className="h-5 w-5" />
            </button>
          </div>
        )}
      </div>
    </div>
  );
};

export default RoutineCard;
