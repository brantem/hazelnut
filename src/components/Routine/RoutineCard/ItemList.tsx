import { useState, useCallback } from 'react';
import {
  DndContext,
  closestCenter,
  MouseSensor,
  TouchSensor,
  KeyboardSensor,
  useSensor,
  useSensors,
  UniqueIdentifier,
} from '@dnd-kit/core';
import {
  arrayMove,
  SortableContext,
  useSortable,
  sortableKeyboardCoordinates,
  verticalListSortingStrategy,
} from '@dnd-kit/sortable';
import { restrictToVerticalAxis, restrictToParentElement } from '@dnd-kit/modifiers';
import { CSS } from '@dnd-kit/utilities';
import clsx from 'clsx';

import Checkbox from 'components/Checkbox';

import { useItemsStore, useRoutinesStore, useHistoriesStore } from 'lib/stores';
import { Routine } from 'types/routine';
import { Item as _Item } from 'types/item';

type ItemProps = {
  routine: Routine;
  item: _Item;
  isSortable?: boolean;
};

const Item = ({ routine, item, isSortable }: ItemProps) => {
  const isDone = useHistoriesStore(
    useCallback((state) => state.getIsDone(routine.id, item.id, true), [routine.id, item.id]),
  );
  const save = useHistoriesStore((state) => state.save);
  const { attributes, listeners, setNodeRef, setActivatorNodeRef, transform, transition } = useSortable({
    id: item.id,
  });

  return (
    <li
      ref={setNodeRef}
      className="flex h-8 w-full items-center justify-between space-x-2 pr-1"
      style={{ transform: CSS.Transform.toString(transform), transition }}
      {...attributes}
    >
      {isSortable && (
        <button
          className="-ml-1.5 flex-grow-0 p-1.5"
          ref={setActivatorNodeRef}
          {...listeners}
          data-testid="routine-item-handle"
        >
          <svg viewBox="0 0 20 20" width="12" className="h-4 w-4 text-neutral-500" fill="currentColor">
            <path d="M7 2a2 2 0 1 0 .001 4.001A2 2 0 0 0 7 2zm0 6a2 2 0 1 0 .001 4.001A2 2 0 0 0 7 8zm0 6a2 2 0 1 0 .001 4.001A2 2 0 0 0 7 14zm6-8a2 2 0 1 0-.001-4.001A2 2 0 0 0 13 6zm0 2a2 2 0 1 0 .001 4.001A2 2 0 0 0 13 8zm0 6a2 2 0 1 0 .001 4.001A2 2 0 0 0 13 14z"></path>
          </svg>
        </button>
      )}

      <div className={clsx('w-full', isSortable && 'max-w-[calc(100%-theme(spacing.8))]')}>
        <Checkbox
          label={item.title}
          name={routine.id + '-' + item.id}
          color={routine.color}
          checked={isDone}
          onChange={(e) => save(routine, item, e.target.checked)}
        />
      </div>
    </li>
  );
};

type ItemListProps = {
  routine: Routine;
  isSortable?: boolean;
};

const ItemList = ({ routine, isSortable }: ItemListProps) => {
  const edit = useRoutinesStore((state) => state.edit);
  const items = useItemsStore(useCallback((state) => state.getItemsByIds(routine.itemIds), [routine.itemIds]));
  const sensors = useSensors(
    useSensor(MouseSensor),
    useSensor(TouchSensor),
    useSensor(KeyboardSensor, { coordinateGetter: sortableKeyboardCoordinates }),
  );

  const [itemId, setItemId] = useState<UniqueIdentifier | null>(null);
  const activeIndex = items.findIndex((item) => item.id === itemId);

  if (!items.length) return null;

  return (
    <DndContext
      sensors={sensors}
      collisionDetection={closestCenter}
      onDragStart={({ active }) => active && setItemId(active.id)}
      onDragEnd={({ over }) => {
        setItemId(null);
        if (!over) return;
        const overIndex = items.findIndex((item) => item.id === over.id);
        if (activeIndex === overIndex) return;
        edit(routine.id, { itemIds: arrayMove(items.map((item) => item.id), activeIndex, overIndex) }); // prettier-ignore
      }}
      onDragCancel={() => setItemId(null)}
      modifiers={[restrictToVerticalAxis, restrictToParentElement]}
    >
      <SortableContext items={items} strategy={verticalListSortingStrategy}>
        <ol className="space-y-1 pt-1" data-testid="routine-card-items">
          {items.map((item) => (
            <Item key={item.id} routine={routine} item={item} isSortable={isSortable} />
          ))}
        </ol>
      </SortableContext>
    </DndContext>
  );
};

export default ItemList;
