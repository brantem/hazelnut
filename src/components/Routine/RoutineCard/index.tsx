import { EllipsisHorizontalIcon, ChevronUpIcon } from '@heroicons/react/20/solid';
import clsx from 'clsx';

import Card from 'components/Card';
import ItemList from 'components/Routine/RoutineCard/ItemList';

import { useRoutinesStore } from 'lib/stores';
import type { Routine } from 'types/routine';
import * as constants from 'data/constants';
import { useModal } from 'lib/hooks';

type RoutineProps = {
  routine: Routine;
  showAction?: boolean;
  isItemSortable?: boolean;
};

const RoutineCard = ({ routine, showAction, isItemSortable = false }: RoutineProps) => {
  const { setRoutine, edit } = useRoutinesStore((state) => ({ setRoutine: state.setRoutine, edit: state.edit }));
  const saveItemsModal = useModal(constants.modals.saveItemsToRoutine);
  const settingsModal = useModal(constants.modals.routineSettings);

  return (
    <Card
      color={routine.color}
      title={
        <>
          <span>{routine.title}</span>
          {routine.time && (
            <span className={`text-sm font-medium text-${routine.color}-400 ml-2 flex-shrink-0 tabular-nums`}>
              {routine.time}
            </span>
          )}
        </>
      }
      actions={[
        {
          children: 'Items',
          onClick: () => {
            setRoutine(routine);
            saveItemsModal.show();
          },
          testId: 'routine-card-save-items',
          skip: !showAction,
        },
        {
          children: <EllipsisHorizontalIcon className="h-5 w-5" />,
          onClick: () => {
            setRoutine(routine);
            settingsModal.show();
          },
          testId: 'routine-card-settings',
          skip: !showAction,
        },
        {
          children: <ChevronUpIcon className={clsx('h-5 w-5', routine.minimized && 'rotate-180')} />,
          onClick: () => edit(routine.id, { minimized: !routine.minimized }),
          testId: 'routine-card-minimize',
        },
      ]}
      data-testid="routine-card"
    >
      {!routine.minimized && <ItemList routine={routine} isSortable={isItemSortable} />}
    </Card>
  );
};

export default RoutineCard;
