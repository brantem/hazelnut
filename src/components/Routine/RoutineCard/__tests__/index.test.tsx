import { render, screen, act, renderHook } from '@testing-library/react';
import '@testing-library/jest-dom';

import RoutineCard from 'components/Routine/RoutineCard';

import { useRoutineStore, useItemsStore } from 'lib/stores';
import { Routine } from 'types/routine';

const routine: Routine = {
  id: 'routine-1',
  title: 'Routine 1',
  color: 'red',
  days: ['MONDAY'],
  time: '00:00',
  itemIds: ['item-1'],
};

describe('RoutineCard', () => {
  beforeEach(() => {
    const items = renderHook(() => useItemsStore());
    act(() => items.result.current.add('group-1', { id: 'item-1', title: 'Item 1' } as Routine));
  });

  afterEach(() => {
    const items = renderHook(() => useItemsStore());
    act(() => items.result.current.remove('item-1'));
  });

  it('should render successfully', () => {
    const { result } = renderHook(() => useRoutineStore());
    const showSaveItems = vi.spyOn(result.current, 'showSaveItems').mockImplementation(() => {});
    const showSettings = vi.spyOn(result.current, 'showSettings').mockImplementation(() => {});

    const { container } = render(<RoutineCard routine={routine} showAction />);

    expect(container).toMatchSnapshot();

    expect(container).toMatchSnapshot();
    expect(screen.getAllByTestId('routine-card-items-item')).toHaveLength(1);

    act(() => screen.getByTestId('routine-card-save-items').click());
    expect(showSaveItems).toHaveBeenCalledWith(routine);

    act(() => screen.getByTestId('routine-card-settings').click());
    expect(showSettings).toHaveBeenCalledWith(routine);
  });

  it("shouldn't show action button", () => {
    const { container } = render(<RoutineCard routine={routine} />);

    expect(container).toMatchSnapshot();
    expect(screen.queryByTestId('routine-card-save-items')).not.toBeInTheDocument();
    expect(screen.queryByTestId('routine-card-settings')).not.toBeInTheDocument();
  });
});
