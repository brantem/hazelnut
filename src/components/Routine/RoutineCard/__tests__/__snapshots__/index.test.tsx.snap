// Vitest Snapshot v1

exports[`RoutineCard > should render successfully 1`] = `
<div>
  <div
    class="px-4 py-3 bg-red-50"
    data-testid="routine-card"
  >
    <div
      class="flex h-7 items-center justify-between space-x-3"
    >
      <div
        class="flex max-w-[440px] items-center space-x-3"
      >
        <h3
          class="text-sm font-semibold uppercase text-red-600 truncate"
        >
          Routine 1
           
        </h3>
        <span
          class="text-sm font-medium text-red-400 ml-2 flex-shrink-0 tabular-nums"
        >
          00:00
        </span>
      </div>
      <div
        class="flex flex-shrink-0 items-center space-x-1"
      >
        <button
          class="rounded-md px-2 py-1 text-sm hover:bg-red-100 flex-shrink-0"
          data-testid="routine-card-save-items"
        >
          Items
        </button>
        <button
          class="rounded-md p-1 hover:bg-red-100"
          data-testid="routine-card-settings"
        >
          <svg
            aria-hidden="true"
            class="h-5 w-5"
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10 3a1.5 1.5 0 110 3 1.5 1.5 0 010-3zM10 8.5a1.5 1.5 0 110 3 1.5 1.5 0 010-3zM11.5 15.5a1.5 1.5 0 10-3 0 1.5 1.5 0 003 0z"
            />
          </svg>
        </button>
      </div>
    </div>
    <ol
      class="space-y-1 pt-2 pb-1"
      data-testid="routine-card-items"
    >
      <li
        class="flex items-center justify-between space-x-3"
        data-testid="routine-card-items-item"
      >
        <div
          class="flex w-full items-center"
        >
          <input
            aria-checked="false"
            class="h-4 w-4 rounded border-red-300 text-red-600 focus:ring-red-500"
            id="routine-1-item-1"
            name="routine-1-item-1"
            type="checkbox"
          />
          <label
            class="ml-3 flex-1 truncate font-medium text-neutral-700"
            for="routine-1-item-1"
          >
            Item 1
          </label>
        </div>
      </li>
    </ol>
  </div>
</div>
`;

exports[`RoutineCard > should render successfully 2`] = `
<div>
  <div
    class="px-4 py-3 bg-red-50"
    data-testid="routine-card"
  >
    <div
      class="flex h-7 items-center justify-between space-x-3"
    >
      <div
        class="flex max-w-[440px] items-center space-x-3"
      >
        <h3
          class="text-sm font-semibold uppercase text-red-600 truncate"
        >
          Routine 1
           
        </h3>
        <span
          class="text-sm font-medium text-red-400 ml-2 flex-shrink-0 tabular-nums"
        >
          00:00
        </span>
      </div>
      <div
        class="flex flex-shrink-0 items-center space-x-1"
      >
        <button
          class="rounded-md px-2 py-1 text-sm hover:bg-red-100 flex-shrink-0"
          data-testid="routine-card-save-items"
        >
          Items
        </button>
        <button
          class="rounded-md p-1 hover:bg-red-100"
          data-testid="routine-card-settings"
        >
          <svg
            aria-hidden="true"
            class="h-5 w-5"
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10 3a1.5 1.5 0 110 3 1.5 1.5 0 010-3zM10 8.5a1.5 1.5 0 110 3 1.5 1.5 0 010-3zM11.5 15.5a1.5 1.5 0 10-3 0 1.5 1.5 0 003 0z"
            />
          </svg>
        </button>
      </div>
    </div>
    <ol
      class="space-y-1 pt-2 pb-1"
      data-testid="routine-card-items"
    >
      <li
        class="flex items-center justify-between space-x-3"
        data-testid="routine-card-items-item"
      >
        <div
          class="flex w-full items-center"
        >
          <input
            aria-checked="false"
            class="h-4 w-4 rounded border-red-300 text-red-600 focus:ring-red-500"
            id="routine-1-item-1"
            name="routine-1-item-1"
            type="checkbox"
          />
          <label
            class="ml-3 flex-1 truncate font-medium text-neutral-700"
            for="routine-1-item-1"
          >
            Item 1
          </label>
        </div>
      </li>
    </ol>
  </div>
</div>
`;

exports[`RoutineCard > shouldn't show action button 1`] = `
<div>
  <div
    class="px-4 py-3 bg-red-50"
    data-testid="routine-card"
  >
    <div
      class="flex h-7 items-center justify-between space-x-3"
    >
      <div
        class="flex max-w-[440px] items-center space-x-3"
      >
        <h3
          class="text-sm font-semibold uppercase text-red-600 truncate"
        >
          Routine 1
           
        </h3>
        <span
          class="text-sm font-medium text-red-400 ml-2 flex-shrink-0 tabular-nums"
        >
          00:00
        </span>
      </div>
    </div>
    <ol
      class="space-y-1 pt-2 pb-1"
      data-testid="routine-card-items"
    >
      <li
        class="flex items-center justify-between space-x-3"
        data-testid="routine-card-items-item"
      >
        <div
          class="flex w-full items-center"
        >
          <input
            aria-checked="false"
            class="h-4 w-4 rounded border-red-300 text-red-600 focus:ring-red-500"
            id="routine-1-item-1"
            name="routine-1-item-1"
            type="checkbox"
          />
          <label
            class="ml-3 flex-1 truncate font-medium text-neutral-700"
            for="routine-1-item-1"
          >
            Item 1
          </label>
        </div>
      </li>
    </ol>
  </div>
</div>
`;
